### PR TITLE
More flexible and deduplicated serialization

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

closes: #130, #145, #461

depends on #447 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (master)
- [ ] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: ec/src/models/short_weierstrass/affine.rs [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+    }
+}
+
+impl<P: SWCurveConfig> Neg for Affine<P> {
+    type Output = Self;
+
+    /// If `self.is_zero()`, returns `self` (`== Self::zero()`).
+    /// Else, returns `(x, -y)`, where `self = (x, y)`.
+    #[inline]
+    fn neg(mut self) -> Self {
```
Comment (weikengchen) [inline ec/src/models/short_weierstrass/affine.rs:245]: Just a confirmation that we are on the same page that if one does `negq = q.neg()`, an implicit copy happens, and q will remain unchanged, right? @confuseSUN mentioned this to me a few hours ago.

**Review by weikengchen — APPROVED** [review]
I went through all the files.

File: ec/src/models/short_weierstrass/affine.rs (hunk 1 above)
Comment (Pratyush) [reply ec/src/models/short_weierstrass/affine.rs:245]: Yes, it’s not a `neg_in_place` (and in fact there’s no such operation)

**Comment by Pratyush** [comment]
@weikengchen There are a few changes here.

* Serialization for Short Weierstrass curves now sets the flags to 0 if the y-coordinate is positive, 1 << 6 if it's a point at infinity, and 1 << 7 if the y-coordinate is negative.
* Serialization for Twisted Edwards curves now sets the flags to 0 if the y-coordinate is positive, and 1 if it's negative. 
I will add this information to the CHANGELOG.

I intend to add a PR that allows users to specify custom serialization/deserialization routines. This is useful, for example, for decoding bls12_381 points, since the standardized format for those curves differs from our standardized format. It would also be useful for users of 0.3 that need to maintain compatibility with the old way of serialization/deserialization.

We should also prepare a migration guide that explains how to go about upgrading to 0.4 once we've finished the refactors.

**Comment by Pratyush** [comment]
I have updated the CHANGELOG now.

**Pull request merged**